pkg/synchronizer/opstate.go

```diff
@@ -0,0 +1,198 @@
+}
+
+// given an IMSI, return the simCard that has that IMSI, or nil if none exists
+func (s *Synchronizer) getSimCardByImsi(site *Site, imsi uint64) *SimCard {
+	for _, sim := range site.SimCard {
+		if sim.Imsi != nil && *sim.Imsi == imsi {
+			return sim
+		}
+	}
+	return nil
+}
+
+// given an IMSI, return the device that has a simcart that has that IMSI, or nil if none exists
+func (s *Synchronizer) getDeviceByImsi(site *Site, imsi uint64) *Device {
+	sim := s.getSimCardByImsi(site, imsi)
+	if sim == nil {
+		return nil
+	}
+	for _, dev := range site.Device {
+		if *dev.SimCard == *sim.SimId {
+			return dev
+		}
+	}
+	return nil
+}
+
+// get a prometheus metrics fetcher, caching previously-created fetchers
+func (s *Synchronizer) getFetcher(endpoint string) (*metrics.Fetcher, error) {
+	mf, okay := s.prometheus[endpoint]
+	if okay {
+		return mf, nil
+	}
+
+	mf, err := metrics.NewFetcher(endpoint)
+	if err != nil {
+		return nil, err
+	}
+	s.prometheus[endpoint] = mf
+	return mf, nil
+}
+
+// update Device.IPAddress opstate
+func (s *Synchronizer) updateDeviceIPAddress(config *gnmi.ConfigForest, entID string, siteID string, imsi uint64, mobileIP string, connected float64) {
+	// Hold the lock while updating the config tree
+	config.Mu.Lock()
+	defer config.Mu.Unlock()
+
+	entRoot, okay := config.Configs[entID]
+	if !okay {
+		// Rare corner case -- the enterprise has disappeared. Maybe some other thread deleted it
+		log.Warnf("Failed to get enterprise %s", entID)
+		return
+	}
+	enterprise := entRoot.(*RootDevice)
+
+	site, okay := enterprise.Site[siteID]
+	if !okay {
+		// Rare corner case -- the site has disappeared. Maybe some other thread deleted it
+		log.Warnf("Failed to get site %s", siteID)
+		return
+	}
+
+	device := s.getDeviceByImsi(site, imsi)
+	if device == nil {
+		// Can happen if the device is reported in prometheus, but does not exist in
+		// ROC. Rare in real life, but plausible in test infrastructure.
+		// Message is currently disabled, as it is fairly noisy.
+		//log.Warnf("Failed to find device %v in site %s of enterprise %s", imsi, siteID, entID)
+		return
+	}
+
+	if device.State == nil {
+		device.State = &DeviceState{}
+	}
+	device.State.IpAddress = aStr(mobileIP)
+
+	if connected != 0 {
+		device.State.Connected = aStr("Yes")
+	} else {
+		device.State.Connected = aStr("No")
+	}
+}
+
+// poll promtheus for metrics, and add them to the config tree
+func (s *Synchronizer) pollPrometheus(config *gnmi.ConfigForest, prom *prometheusInfo) error {
+	log.Debugf("Polling prometheus site %s enterprise %s endpoint %s", prom.siteID, prom.entID, prom.endpoint)
+
+	fetcher, err := s.getFetcher(prom.endpoint)
+	if err != nil {
+		return err
+	}
+	v, err := fetcher.GetVector("subscribers_info")
+	if err != nil {
+		return err
+	}
+
+	// vector is []*Sample
+	// sample is {Metric Value:SampleValue Time}
+	// metric is a labelset
+	// labelsite is map[labelname]labelvalue
+	// samplevalue is float64
+
+	for _, sample := range v {
```

Comment:
I'm not clear how big this v array could be - is it all the past samples in the time dimension - you keep looping until you get to the end of the range - are they going to be ordered in time? maybe you should just take the last one?

Reply:
It will only return the set of values "at the current time". The exact behavior depends on what the prometheus producer (i.e. sd-core) does. If the producer continues to report old information with a current timestamp, then we would continue to see that old information. sd-core used to have this bug, but I think they've fixed it.

Reply:
OK